scripts/executors/src/deprecate-react-components-preview-packages.spec.ts

```diff
@@ -0,0 +1,152 @@
+const mockReadJsonFile = jest.spyOn(devkit, 'readJsonFile');
+const mockJoin = jest.spyOn(path, 'join');
+const mockGetAllPackageInfo = jest.spyOn(monorepoScripts, 'getAllPackageInfo');
+
+describe('deprecatePackage', () => {
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('should execute npm deprecate command', () => {
+    mockExecSync.mockImplementation(() => '');
+
+    deprecatePackage('test-package', 'npm-token');
+
+    expect(mockExecSync).toHaveBeenCalledWith(
+      'npm deprecate test-package "Deprecated in favor of stable release" --registry https://registry.npmjs.org/ --//registry.npmjs.org/:_authToken=npm-token',
+      { stdio: 'inherit' },
+    );
+  });
+
+  it('should log error when deprecation fails', () => {
+    mockExecSync.mockImplementationOnce(() => {
+      throw new Error('Deprecation failed');
+    });
+
+    expect(() => deprecatePackage('test-package', 'npm-token')).toThrowError(
+      'Failed to deprecate "test-package" package',
+    );
+  });
+});
+
+describe('getPackagesToDeprecate', () => {
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('should return correct preview packages to deprecate', () => {
+    setup({
+      // package to deprecate
+      'package-a': {
+        changeFile: { type: 'minor' },
+        version: '9.0.0-alpha.0',
+      },
+      // ignore, as there is no change file for the package
+      'package-b': {
+        changeFile: null,
+        version: '9.4.1',
+      },
+      // package to deprecate
+      'package-c': {
+        changeFile: { type: 'minor' },
+        version: '9.0.0-alpha.0',
+      },
+      // ignore, as it has a `-preview` suffix
+      'package-d-preview': {
+        changeFile: { type: 'minor' },
+        version: '9.0.0-alpha.0',
+      },
+      // ignore, as the package version is not `9.0.0-alpha.0`
+      'package-e': {
+        changeFile: { type: 'minor' },
+        version: '9.1.2',
+      },
+    });
+
+    const result = getPackagesToDeprecate();
+    expect(result).toEqual(['@fluentui/package-a', '@fluentui/package-c']);
+  });
+
+  type PreviewPackageConfig = {
+    changeFile: { type: ChangeType } | null;
+    version: string;
+  };
+
+  function setup(packages: Record<string, PreviewPackageConfig> = {}) {
+    const allPackages = Object.entries(packages).reduce<monorepoScripts.AllPackageInfo>(
+      (acc, [packageName, { version }]) => {
+        const packageNameWithScope = `@fluentui/${packageName}`;
+        acc[packageNameWithScope] = {
+          packagePath: '',
+          packageJson: {
+            name: packageNameWithScope,
+            version,
+            main: 'dist/index.js',
+          },
+          projectConfig: {
+            root: `packages/${packageName}`,
+            name: packageName,
+            tags: ['web', 'vNext'],
+          },
+        };
+
+        return acc;
+      },
+      {},
+    );
+
+    // Mock all packages info
+    mockGetAllPackageInfo.mockImplementation(predicate => {
```

Comment:
hmm this is "re-implementing" the whole logic of the function which makes the test even more brittle (besides other mocks).

I'd suggest to do following (can be a PR follow-up):

> you can take look for inspiration here https://github.com/microsoft/fluentui/blob/master/scripts/generators/src/generate-version-files.ts#L20

- re-do the API in order to be able to implement more robust test ( which is necessary as this will become crucial part of our release process )
- passing change root dir to main program (acquired from args). 
- this will allow us to create `__fixtures__/`(where we can create artificially change-files for our test cases) for our test while running things against actual monorepo graph ( no need for mocking all the logic )

with this setup, only valid mock(required as well) would be for `execSync` as we don't wanna run `npm deprecate` ofc


### `main()`:

```ts
function main() {
  const argv = yargs
    .option('changeFilesRoot', {
      type: 'string',
      description: 'Root folder where change files live (relative to workspace root)',
      default: 'change',
    })
    .option('token', {
      type: 'string',
      description: 'NPM Token',
      demandOption: true,
    }).argv;

  deprecateReactComponentsPreviewPackages({ argv });
}

if (require.main === module) {
  main();
}
```

## `deprecateReactComponentsPreviewPackages`

> this is the only function that we will "e2e" test
>
> NOTE: logic of acquiring the change files ( the factory ) etc would need to be re-adjusted

```ts
export function deprecateReactComponentsPreviewPackages(options: { argv: { token: string; changeFilesRoot: string } }) {
  const { argv } = options;

  try {
    // do stuff here as needed
  } catch (e) {
    logger.error(e);
    logger.error('Failed to deprecate packages');
    process.exit(1);
  }
}
```

Reply:
upd: discussed offline and agreed to use this approach with the one modification - we'll inject `allPackages` information into the function